docs/sources/flow/reference/components/otelcol.processor.attributes.md

```diff
@@ -0,0 +1,582 @@
+* For logs, one of `log_bodies`, `log_severity_texts`, [attribute], [resource], or [library] must be specified with a non-empty value for a valid configuration. The `span_names`, `span_kinds`, `metric_names`, `resource_attributes`, and `services` fields are invalid.
+* For metrics, one of `metric_names`, [resource] must be specified with a valid non-empty value for a valid configuration. The `span_names`, `span_kinds`, `log_bodies`, `log_severity_texts` and `services` fields are invalid.
+
+For `metric_names`, a match occurs if the metric name matches at least one item in the list.
+For `span_kinds`, a match occurs if the span's span kind matches at least one item in this list.
+
+Note: If both `include` and `exclude` are specified, the include properties are checked before the exclude properties.
+
+### regexp block
+
+This block is optional configuration for the `match_type` of `"regexp"`.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`cacheenabled` | `bool` | Determines whether match results are LRU cached. | `false` | no
+`cachemaxnumentries` | `int` | The max number of entries of the LRU cache that stores match results. | `0` | no
+
+Enabling `cacheenabled` could make subsequent matches faster.
+Cache size is unlimited unless `cachemaxnumentries` is also specified.
+
+`cachemaxnumentries` is ignored if `cacheenabled` is false.
+
+### attribute block
+
+This block specifies a list of attributes to match against.
+Only `match_type = "strict"` is allowed if `attribute` is specified.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`key` | `string` | The attribute key. | | yes
+`value` | `any` | The value to match against. | | no
+
+If `value` is not set, any value will match.
+The type of `value` could be a number, a string or a boolean.
+
+### resource block
+
+This block specifies items to match the resources against.
+A match occurs if the input data resources matches at least one `resource` block.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`key` | `string` | The attribute key. | | yes
+`value` | `any` | The value to match against. | | no
+
+If `value` is not set, any value will match.
+The type of `value` could be a number, a string or a boolean.
+
+### library block
+
+This block specifies items to match the implementation library against.
+A match occurs if the span's implementation library matches at least one `library` block.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`name` | `string` | The attribute key. | | yes
+`version` | | The value to match against. | | yes
```

Comment:
`version` is a pointer. Should we include a note in the docs that leaving this unset and setting it to an empty string are not the same thing, and could lead to different outcomes?

Reply:
I don't think it has to be a pointer on our side, right? We can turn it into a pointer when converting to the upstream type.

Reply:
The matching rules for null pointers and empty strings work differently:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/filter/filterconfig/config.go#L214-L226

If `version` is a null pointer, it will match any version. If `version` is an empty string, it will only match a version which is also an empty string.

I will try to add some unit tests to make sure this works. If it doesn't work, then I could probably add an extra bool flag in the river config so that our code sets the version to a null pointer during the conversion.